---
title: "CD8_APLcombined_unstimulated_noise"
author: "Arianne Richard"
date: "April 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '~/Documents/Post_doc/Git/SingleCellAPL2018/scRNAseq/')
```

This file searches for structure or hints of what may make certain cells better responders in the unstimulated condition.

First we load the data, remove uninformative genes, and normalize the data.

```{r}
library(scater)
library(scran)
library(SingleCellExperiment)
library(RColorBrewer)
library(gplots)
source('analysis/dpt_plots.R')
source('analysis/dpt.R')
```


First we will look within the single batch where the unstimulated cells were and look at pseudotime from the original N4 timecourse. We'll first process it as we did previously.

```{r}
sce <- readRDS('data/sce_run2.rds')
drop_filt <- rowSums(counts(sce) > 0) >= (ncol(sce) * 0.05)
sce <- sce[drop_filt,]
mean_filt <- rowMeans(counts(sce)) >= 1
sce <- sce[mean_filt,]

sceunind <- sce[,sce$Condition %in% 'unstimulated']
sceunind <- computeSpikeFactors(sceunind, general.use=TRUE)
sceunind <- normalize(sceunind, exprs_values='counts', return_norm_as_exprs=TRUE, log_exprs_offset=1)

sce <- computeSpikeFactors(sce, general.use=TRUE)
sce <- normalize(sce, exprs_values='counts', return_norm_as_exprs=TRUE, log_exprs_offset=1)

```

# Highly variable gene analysis

What are the highly variable genes?

```{r}

sceunind <- sceunind[!isSpike(sceunind),]

df <- data.frame(batch=as.character(sceunind$Sorting.Plate))
mod <- model.matrix(~0+batch, data=df)
alt.fit <- trendVar(sceunind, use.spikes=FALSE, design=mod, span=0.15) 
alt.decomp <- decomposeVar(sceunind, alt.fit, design=mod)
fdat <- rowData(sceunind)
rownames(fdat) <- fdat$GeneID
alt.decomp$symbol <- fdat[rownames(alt.decomp), 'symbol']
top.hvgs <- order(alt.decomp$bio, decreasing=TRUE)
head(alt.decomp[top.hvgs,], 20)

plot(alt.decomp$mean, alt.decomp$total, xlab="Mean log-expression", ylab="Variance")
o <- order(alt.decomp$mean)
lines(alt.decomp$mean[o], alt.decomp$tech[o], col="red", lwd=2)


```

Ordering cells by decreasing biological variability reveals adhesion molecules in the top 5. Itgb2 is CD18, part of LFA-1, which is known to play a role in T cell activation.

In addition, Fos is the only significantly highly variable gene.

Let's look at Fos and the LFA-1 components.

```{r}

expr_box_plot <- function(gene_symbol, sceset, 
                     split_facs=c('unstimulated', 'N4_1h', 'N4_3h', 'N4_6h'), 
                     name_facs=c('unstimulated', 'N4_1h', 'N4_3h', 'N4_6h'), 
                     cols=c('grey', 'gold', 'orange', 'red')){
  fdat <- rowData(sceset)
  rownames(fdat) <- fdat$GeneID
  genen <- rownames(fdat)[fdat$symbol %in% gene_symbol]
  temp <- split(exprs(sceset)[genen,], 
                factor(as.character(colData(sceset)$Condition), levels=split_facs))
  names(temp) <- name_facs
  stripchart(temp, method='jitter', vertical=TRUE, pch=20, col=cols, las=2, 
             xlim=c(0.5,length(temp)+0.5), 
             ylab=bquote('log'[2] ~ .(gene_symbol) ~ 'expression'))
  boxplot(temp, outline=FALSE, add=T, col=rgb(0,0,0, alpha=0), axes=F)
}

par(mfrow=c(1,3), mgp=c(2,0.5,0), mar=c(6,3.5,1,0.2), bty='l', cex.lab=1.5, cex.axis=1)
expr_box_plot('Itgal', sceset=sce)
expr_box_plot('Itgb2', sceset=sce)
expr_box_plot('Fos', sceset=sce)

pdf('data/APL_Unstim_LFA1exp.pdf', height=4, width=6)
par(mfrow=c(1,2), mgp=c(2,0.5,0), mar=c(6,3.5,1,0.2), bty='l', cex.lab=1.5, cex.axis=1)
expr_box_plot('Itgal', sceset=sce)
expr_box_plot('Itgb2', sceset=sce)
dev.off()

pdf('data/APL_Unstim_Fosexp.pdf', height=4, width=6)
par(mfrow=c(1,2), mgp=c(2,0.5,0), mar=c(6,3.5,1,0.2), bty='l', cex.lab=1.5, cex.axis=1)
expr_box_plot('Fos', sceset=sce)
dev.off()
```

# Examining the "progressive" unstimulated cells

In our pseudotime analysis, there were a few unstimuated cells that crept into the "early activation" cluster with the majority of cells stimulated for 1 hour. Let's see what is special about these cells. 

We first need to batch correct the whole timecourse dataset as we did previously.

```{r}

library(sva)
mod <- model.matrix(~factor(as.character(sce$Condition)))
combat_sce <- ComBat(dat=exprs(sce), batch=colData(sce)$Sorting.Plate,
                     mod=mod, par.prior=TRUE, prior.plots=FALSE)
exprs(sce) <- combat_sce
sce <- sce[, sce$Condition %in% c('unstimulated', 'N4_1h', 'N4_3h', 'N4_6h')]

cond <- as.character(sce$Condition)
cond[cond %in% 'unstimulated'] <- '0h'
cond[cond %in% 'N4_1h'] <- '1h'
cond[cond %in% 'N4_3h'] <- '3h'
cond[cond %in% 'N4_6h'] <- '6h'
sce$Condition <- cond

```

Now we can look at pseudotime etc.

```{r}
sce <- sce[!isSpike(sce),]

library(destiny)

resTC <- quick_dpt(sce, conditions=c('0h', '1h', '3h', '6h'), col=c('grey', 'gold', 'orange', 'red'), base_condition='0h', stim_condition='6h')

resTC[[1]]

library(classInt)
set.seed(100)
clusts <- findCols(classIntervals(resTC$pt$DPT, 3, style='jenks'))

col_cat <- c('resting', 'early activation', 'late activation')[as.numeric(clusts)]
col_cat <- factor(col_cat, levels=c('resting', 'early activation', 'late activation'))

plot_dpt_ACR_all_colcat(ts=resTC$ts, pt=resTC$pt, branches=c(1L), path_col=c('black'),
                        col_cat = col_cat, cat_cols=c('darkgrey', 'lightslateblue', 'slateblue4'), 
                        col_name='Clusters',
                        x = 1L, y = 2L, w_width = 0.4)

```

So the question is what is unique about the cells that have ended up in the early activation cluster despite being unstimulated.

Let's look at which plate they were collected on in case the later plates had a mild level of activation due to slight warming.

```{r}

col_cat <- sce$Sorting.Plate

plot_dpt_ACR_all_colcat(ts=resTC$ts, pt=resTC$pt, branches=c(1L), path_col=c('black'),
                        col_cat = col_cat, cat_cols=c('purple', 'darkgreen', 'blue', 'magenta'), 
                        col_name='Plates',
                        x = 1L, y = 2L, w_width = 0.4)

```

There is a tendency for plate 4 unstimulated cells to be a bit more progressive (4 of the 6 progressive cells), but the correlation isn't spectacular because the other 2 are from plate 1. It may be that they warmed up a bit during the sorting time and this allowed them to exhibit a slight stress response.

# Differential expression analysis

Regardless, we want to ask what is special about these progressive cells.  So we'll first try differential expression analysis.

```{r}
unind <- which(sce$Condition %in% '0h')
sceunind <- sce[,unind]
progressive_cells <- colnames(sceunind)
clusts_un <- clusts[unind]
progressive_cells <- progressive_cells[which(clusts_un %in% 2)]

library(edgeR)
df <- data.frame(batch=as.character(sceunind$Sorting.Plate), 
                 clust=as.character(clusts_un))
mod <- model.matrix(~0+clust+batch, data=df)
z_s <- convertTo(sceunind, type="edgeR")
y_s <- estimateDisp(z_s, design=mod)
fit <- glmFit(z_s, design=mod, dispersion=y_s$tagwise.dispersion, prior.count=1)
res <- glmLRT(fit, contrast=c(-1,1,0,0,0))
bestun <- data.frame(topTags(res, n=nrow(z_s)))
bestun$symbol <- as.character(fdat[rownames(bestun),'symbol'])
head(bestun, 20)

boxplot(split(exprs(sceunind)[rownames(bestun)[1], ], factor(clusts_un)), ylab=bestun[1, 'symbol'])

```

Nothing is significant.

What if we consider pseudotime designation a continuous variable?

```{r}
df <- data.frame(batch=as.character(sceunind$Sorting.Plate), 
                 dpt=resTC$pt$DPT[unind])
mod <- model.matrix(~0+dpt+batch, data=df)
z_s <- convertTo(sceunind, type="edgeR")
y_s <- estimateDisp(z_s, design=mod)
fit <- glmFit(z_s, design=mod, dispersion=y_s$tagwise.dispersion, prior.count=1)
res <- glmLRT(fit, coef='dpt')
bestun <- data.frame(topTags(res, n=nrow(z_s)))
bestun$symbol <- as.character(fdat[rownames(bestun),'symbol'])
head(bestun, 20)

plot(resTC$pt$DPT[unind], 
     exprs(sceunind)[rownames(bestun)[1], ], 
     ylab=bestun[1, 'symbol'],
     xlab='dpt', 
     col=c('black', 'red')[clusts_un])

```

The only significant association is with Ift52. Ift52 is a gene of potential interest as it is actually associated with cilia. However, while it is associated with some of those advanced cells, it is certainly not deterministic in defining cells that have entered the early activation cluster.

# Principal components analysis

What happens if we try looking at PCs for these cells?

```{r}
pcs <- prcomp(t(as.matrix(exprs(sceunind))), scale=TRUE)
pairs(pcs$x[,c(1:4)], pch=16, cex.lab=1.2, col=c('black', 'red')[clusts_un])

pdf('data/APL_TC_Unstim_PCs.pdf', height=6, width=6)
pairs(pcs$x[,c(1:4)], pch=16, cex.lab=1.2, col=c('black', 'red')[clusts_un])
dev.off()

```

So these "progressive" cells appear to be the most divergent in PCA but to have no specific gene expression alterations.

# tSNE analysis

Let's look at whether there is any structure by tSNE in the unstimulated cells.

```{r}

library(Rtsne)
tn <- Rtsne(t(exprs(sceunind)), perplexity=10)
plot(tn$Y[,1], tn$Y[,2], col=c('black', 'red')[clusts_un])

```

There is basically no structure here either.

# Examination of activation markers

For completeness, let's do a hypothesis-driven check for signs of activation.

```{r}
genes <- c('Nr4a1', 'Il2ra', 'Myc', 'Gzmb', 'Cd44', 'Cd69', 'Cd8a', 'Ptpn6')
par(mfrow=c(2,4))
for (i in 1:length(genes)){
  geneID <- rownames(fdat)[fdat$symbol %in% genes[i]]
  ct <- cor.test(resTC$pt$DPT[unind], exprs(sceunind)[geneID, ])
  plot(resTC$pt$DPT[unind], 
       exprs(sceunind[geneID,]), 
       col=c('black', 'red')[clusts_un],
       pch=16,
       xlab='pseudotime',
       ylab=paste0('log RNA expression ', genes[i]), 
       main=paste0('r ', as.character(signif(ct$estimate, 2)), ', p ', as.character(signif(ct$p.value, 2))))
}

pdf('data/APL_TC_Unstim_selected_genes.pdf', height=6, width=10)
par(mfrow=c(2,4), mgp=c(2,0.5,0), bty='l', cex.lab=1.5, cex.axis=1)
for (i in 1:length(genes)){
  geneID <- rownames(fdat)[fdat$symbol %in% genes[i]]
  plot(resTC$pt$DPT[unind], 
       exprs(sceunind[geneID,]), 
       col=c('black', 'red')[clusts_un],
       pch=16,
       main=genes[i],
       xlab='pseudotime',
       ylab='log RNA expression')
}
dev.off()

proteins <- c('Characteristics.CD25.measurement..log10..', 'Characteristics.CD69.measurement..log10..', 'Characteristics.CD44.measurement..log10..', 'Characteristics.CD62L.measurement..log10..')

par(mfrow=c(2,4), mgp=c(2,0.5,0), bty='l', cex.lab=1.5, cex.axis=1)
for(i in 1:length(proteins)){
  ct <- cor.test(resTC$pt$DPT[unind], 
                 as.numeric(as.character(colData(sceunind)[, proteins[i]])))
  plot(resTC$pt$DPT[unind], 
       as.numeric(as.character(colData(sceunind)[, proteins[i]])),
       xlab='DPT', 
       ylab=paste0('log10 ', strsplit(proteins[i], '.', fixed=TRUE)[[1]][2]), 
       col=c('black', 'red')[clusts_un],
       main=paste0('r ', as.character(signif(ct$estimate, 2)), 
                   ', p ', as.character(signif(ct$p.value, 2))),
       pch=16)
}

pdf('data/APL_TC_Unstim_DPTproteins.pdf', height=6, width=6)
par(mfrow=c(2,2), mgp=c(2,0.5,0), bty='l', cex.lab=1.5, cex.axis=1)
for(i in 1:length(proteins)){
  plot(resTC$pt$DPT[unind], 
       as.numeric(as.character(colData(sce)[unind, proteins[i]])),
       xlab='DPT', 
       ylab='log10 protein expression',
       main=strsplit(proteins[i], '.', fix=TRUE)[[1]][2],
       col=c('black', 'red')[clusts_un], pch=16)
}
dev.off()

```

# Clustering

And let's cluster the cells to see if our progressive cells are outlying.

```{r}
sceunind_var <- sceunind[top.hvgs[1:100],]
## note - these genes were calculated using the dataset with batch as an uninteresting variance-contributing factor but before any batch correction and with normalization factors centered on the unstimulated cells only; we are now taking the unstimulated cells from the full normalized and batch-corrected dataset and filtering for these genes
library(pheatmap)
clust_df <- data.frame(clusters=c('resting', 'early activation')[clusts_un])
rownames(clust_df) <- colnames(sceunind_var)
pheatmap(exprs(sceunind_var), annotation_col=clust_df, show_rownames=FALSE, show_colnames=FALSE)

pdf('data/APL_TC_Unstim_HVG_heaptmap.pdf', height=5, width=5)
pheatmap(exprs(sceunind_var), annotation_col=clust_df, show_rownames=FALSE, show_colnames=FALSE)
dev.off()

## and a sanity check for code above
boxplot(split(exprs(sce)[rownames(fdat[fdat$symbol %in% 'Cd69',]),unind], c('resting', 'early activation')[clusts_un]))

```

```{r}
sessionInfo()
```