---
title: "CD8_APLseparated_analyses_2017"
author: "Arianne Richard"
date: "March 2018"
output: html_document
---

This script runs through analyses of the altered affinity ligand stimulation data from a single-cell sequencing experiment looking at early activation of CD8+ T cells - this data was collected from experiments run on 20160727 and 20161026.  These analyses treat each dataset separately.

# Loading the data and filtering genes

First we'll load the data and filter uninformative genes.

```{r}
library(scater)
library(SingleCellExperiment)
library(scran)
library(RColorBrewer)
library(gplots)
library(dpt)
source('dpt_plots.R')
library(pheatmap)

sce1 <- readRDS('../data/sce_run1.rds')
sce2 <- readRDS('../data/sce_run2.rds')

## filter to remove non-expressed genes -- require detection in at least 5% of cells in each batch
drop_filt <- rowSums(counts(sce1) > 0) >= (ncol(sce1) * 0.05)
sce1 <- sce1[drop_filt,]
mean_filt <- rowMeans(counts(sce1)) >= 1
sce1 <- sce1[mean_filt,]

drop_filt <- rowSums(counts(sce2) > 0) >= (ncol(sce2) * 0.05)
sce2 <- sce2[drop_filt,]
mean_filt <- rowMeans(counts(sce2)) >= 1
sce2 <- sce2[mean_filt,]

## calculate spike factors
sce1 <- computeSpikeFactors(sce1, general.use=TRUE)
sce2 <- computeSpikeFactors(sce2, general.use=TRUE)

## normalise the data
sce1 <- normalize(sce1, exprs_values='counts', return_norm_as_exprs=TRUE, log_exprs_offset=1)
sce2 <- normalize(sce2, exprs_values='counts', return_norm_as_exprs=TRUE, log_exprs_offset=1)

```

We know from looking at the time-course analysis that we need to batch correct run 2, so let's do that.

# Batch effect correction

```{r}
library(sva)
mod <- model.matrix(~factor(as.character(sce2$Condition)))
combat_sce <- ComBat(dat=exprs(sce2), batch=colData(sce2)$Sorting.Plate,
                     mod=mod, par.prior=TRUE, prior.plots=FALSE)
sce2_noBC <- sce2
exprs(sce2) <- combat_sce
```

Let's check if we need to do any plate-based batch correction for run1.

```{r}
plot_splitsce <- function(sce){
  splitsce <- vector('list', 4)
  names(splitsce) <- c('N4_6h', 'T4_6h', 'G4_6h', 'NP68_6h')
  for(i in 1:length(splitsce)){
    splitsce[[i]] <- sce[,sce$Condition %in% names(splitsce)[i]]
  }
  splitpcs <- lapply(splitsce, function(x) {prcomp(t(as.matrix(exprs(x))))})
  par(mfrow=c(2,4))
  for(i in 1:length(splitpcs)){
    sce2 <- splitsce[[i]]
    cols2 <- as.character(colData(sce2)$Sorting.Plate)
    cols2[cols2 %in% 'Plate1'] <- 'blue'
    cols2[cols2 %in% 'Plate2'] <- 'green3'
    cols2[cols2 %in% 'Plate3'] <- 'purple'
    cols2[cols2 %in% 'Plate4'] <- 'red'
    par(mgp=c(2,1,0), bty='l')
    plot(splitpcs[[i]]$x[,1], splitpcs[[i]]$x[,2], xlab='PC1', ylab='PC2', 
         main=names(splitpcs)[i], col=cols2, pch=16, cex.lab=1.2)
  }
}

plot_splitsce(sce1)
```

It looks like there could be some subtle effects visible in N4 and G4 categories, so we'll batch correct to play it safe.

```{r}
sce1$Sorting.Plate <- factor(as.character(sce1$Sorting.Plate))
mod <- model.matrix(~factor(as.character(sce1$Condition)))
combat_sce <- ComBat(dat=exprs(sce1), batch=colData(sce1)$Sorting.Plate,
                     mod=mod, par.prior=TRUE, prior.plots=FALSE)
exprs(sce1) <- combat_sce
```

Now we have sce1 and sce2 normalised and batch-corrected.  We'll remove spikes so they do not interfere with downstream analyses.

```{r}
sce1 <- sce1[!isSpike(sce1),]
sce2 <- sce2[!isSpike(sce2),]
sce2_noBC <- sce2_noBC[!isSpike(sce2_noBC),]
```

# Finding a control gene for RNA flow cytometry

We're going to perform RNA flow cytometry to look at earlier time-points with the reduced affinity ligands, but to do this, we need a reliable control gene.  Let's use the sequencing data to find such a control gene. We'll first focus on sce2, which has data from both the time-course and the altered peptide ligand stimulations.  

We'll limit ourselves to genes with less than 20% drop-out in all conditions, and then look for genes with low variance and minimal association with stimulation condition. We'll use non-batch-corrected values within each plate to calculate variance. We note that there is a mean-variance relationship, but for the purposes of finding an invariant gene, we still want the lowest variance, even if that means we are partially selecting on the mean.

```{r}
scesplit <- split(colnames(sce2), factor(sce2$Condition, 
                                        levels=c('N4_6h', 'T4_6h', 'G4_6h', 'NP68_6h', 'unstimulated', 'N4_1h', 'N4_3h')))
scesplit <- lapply(scesplit, function(x) {counts(sce2)[,x]})
drops <- lapply(scesplit, function(y) {
  rowSums(y > 0) >= (ncol(y) * 0.8)})
drops <- data.frame(drops)
drops_filt <- apply(drops, 1, function(x){length(which(x)) >= ncol(drops)})

sce2_CG <- sce2[drops_filt,]
sce2_CG_noBC <- sce2_noBC[drops_filt,]

scesplit <- split(colnames(sce2_CG_noBC), factor(sce2_CG_noBC$Sorting.Plate))
scesplit <- lapply(scesplit, function(x) {exprs(sce2_CG_noBC)[,x]})
vars <- lapply(scesplit, function(y) {
  apply(y, 1, var)})
vars <- data.frame(vars)
vars_avg <- apply(vars, 1, mean)
fdat <- rowData(sce2_CG_noBC)
rownames(fdat) <- rownames(sce2_CG_noBC)
symbols <- fdat[names(vars_avg), 'symbol']
vars_avg <- data.frame(vars_avg, symbols)
head(vars_avg[order(vars_avg[,1]),], 10)
```

Now we'll look for lack of association with condition by ANOVA.

```{r}
anovas <- apply(exprs(sce2_CG), 1, function(z){
  gene <- z
  condition <- sce2_CG$Condition
  test <- aov(gene~condition, data.frame(gene, condition))
  return(summary(test)[[1]][1,5])
})

vars_avg$anova <- anovas[rownames(vars_avg)]
plot(-log10(vars_avg$anova), vars_avg$vars_avg, xlab='-log10(anova P-value)', ylab='variance')
```

Let's keep the ANOVA p-value below 0.01 and the variance below 2.

```{r}
vars_avg <- vars_avg[(vars_avg$vars_avg < 2) & (vars_avg$anova > 0.01), ]
vars_avg[order(vars_avg$anova, decreasing=T),]
```

Looking at the list in order of increasing variance, we'll avoid pseudogenes, non-coding genes etc, which leaves us with Slc34a2 and Rpl39 in the top of the list.  Let's look at these in both scRNA-seq datasets.

```{r}
boxes_plot <- function(sceset, geneID, mylevels=c('N4_6h', 'T4_6h', 'G4_6h', 'NP68_6h', 'unstimulated', 'N4_1h', 'N4_3h'), ANOVA=TRUE) {
  fdat <- rowData(sceset)
  rownames(fdat) <- rownames(sceset)
  genen <- rownames(fdat[fdat$symbol %in% geneID,])
  temp <- split(exprs(sceset)[genen,],
                factor(as.character(sceset$Condition),
                       levels=mylevels))
  stripchart(temp, method='jitter', vertical=TRUE, pch=20, 
             col=c('red', 'green3', 'blue', 'grey40', 'grey', 'gold', 'orange'), 
             las=2, 
             xlim=c(0.5,length(temp)+0.5), ylim=c(0,12), cex.lab=1.2,
             ylab=bquote('log'[2] ~ .(geneID) ~ 'expression'))
  boxplot(temp, outline=FALSE, add=T, col=rgb(0,0,0, alpha=0), axes=F)
  if(ANOVA) {
    a <- signif(summary(aov(gene~condition, 
                            data.frame(gene=exprs(sceset)[genen,],
                                       condition=factor(sceset$Condition))))[[1]][1,5], 3)
    text(x=length(mylevels)-1, y=11, labels=bquote(ANOVA ~ italic(' p ') ~ .(a)), cex=1.2)
  }
}

par(mfrow=c(2,2), bty='l')
boxes_plot(sce2, 'Slc34a2')
boxes_plot(sce2, 'Rpl39')
boxes_plot(sce1, 'Slc34a2', mylevels=c('N4_6h', 'T4_6h', 'G4_6h', 'NP68_6h'))
boxes_plot(sce1, 'Rpl39', mylevels=c('N4_6h', 'T4_6h', 'G4_6h', 'NP68_6h'))

pdf('../data/Control_genes_spikenorm_filt05_boxplot.pdf', height=8, width=8)
par(mfrow=c(2,2), mgp=c(2,0.5,0), mar=c(5.1,4.1,1.5,0.5), bty='l')
boxes_plot(sce2, 'Slc34a2')
boxes_plot(sce2, 'Rpl39')
boxes_plot(sce1, 'Slc34a2', mylevels=c('N4_6h', 'T4_6h', 'G4_6h', 'NP68_6h'))
boxes_plot(sce1, 'Rpl39', mylevels=c('N4_6h', 'T4_6h', 'G4_6h', 'NP68_6h'))
dev.off()
```

It looks like Slc34a2 would be the best control gene. Rpl39 is not bad, but it is associated with condition to a small extent in the second dataset. 

Update: Sadly, testing the PrimeFlow reagents for Slc34a2 yielded no signal, and so we will use Rpl39 as a reliable control gene.

Since we're going to use RNA flow cytometry to look at Nr4a1 and Fosb in the APL data at early timepoints.  Let's just quickly check how these genes look in our RNAseq at 6 hours in APL conditions.

```{r}
par(mfrow=c(2,2), bty='l')
boxes_plot(sce2, 'Nr4a1', ANOVA=FALSE)
boxes_plot(sce2, 'Fosb', ANOVA=FALSE)
boxes_plot(sce1, 'Nr4a1', ANOVA=FALSE, mylevels=c('N4_6h', 'T4_6h', 'G4_6h', 'NP68_6h'))
boxes_plot(sce1, 'Fosb', ANOVA=FALSE, mylevels=c('N4_6h', 'T4_6h', 'G4_6h', 'NP68_6h'))
```

So Fosb is unique to the burst (not even detected in enough cells in the non-time-course dataset) and Nr4a1 is more varied (decent expression in T4 6h) though definitely highest in N4 1h.

# Protein expression in sequenced cells

Finally, let's look at protein expression by condition.

```{r}

protein_violin_plot <- function(protein, sceset, split_facs, cols){
  jitter <- position_jitter(width = 0.07, height = 0)
  protname <- paste0('Characteristics.', protein, '.measurement..log10..')
  temp <- data.frame(colData(sceset)[,c(protname, 'Condition')])
  temp <- temp[temp$Condition %in% split_facs,]
  temp$Condition <- factor(as.character(temp$Condition), levels=split_facs)
  colnames(temp) <- c('protein', 'condition')
  temp$protein <- as.numeric(as.character(temp$protein))
  temp$condition <- unlist(lapply(strsplit(as.character(temp$condition), '_'), function(x){x[1]}))
  temp_levels <- unlist(lapply(strsplit(as.character(split_facs), '_'), function(x){x[1]}))
  temp$condition <- factor(temp$condition, levels=temp_levels)
  ggplot(temp, aes(condition, protein, fill=factor(condition))) + 
    geom_violin() + 
    scale_fill_manual(values=cols) + 
    guides(fill=FALSE) + 
    geom_point (size=0.5, shape=1, position=jitter) +
    theme_bw() + labs(y=bquote('log'[10] ~ .(protein) ~ 'expression', where=environment())) +
    theme(panel.border = element_blank(), panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"))
}


par(mfrow=c(2,2))
protein_violin_plot('CD69', sceset=sce1, 
            split_facs=c('N4_6h', 'T4_6h', 'G4_6h', 'NP68_6h'), 
            cols=c('red', 'green3', 'blue', 'grey40'))
protein_violin_plot('CD69', sceset=sce2, 
            split_facs=c('N4_6h', 'T4_6h', 'G4_6h', 'NP68_6h'), 
            cols=c('red', 'green3', 'blue', 'grey40'))
protein_violin_plot('CD25', sceset=sce1, 
            split_facs=c('N4_6h', 'T4_6h', 'G4_6h', 'NP68_6h'), 
            cols=c('red', 'green3', 'blue', 'grey40'))
protein_violin_plot('CD25', sceset=sce2, 
            split_facs=c('N4_6h', 'T4_6h', 'G4_6h', 'NP68_6h'), 
            cols=c('red', 'green3', 'blue', 'grey40'))


pdf(paste0('../data/CD69_protein_APL1_spikenorm_filt05_violinplot.pdf'), height=2.5, width=2)
protein_violin_plot('CD69', sceset=sce1, 
            split_facs=c('N4_6h', 'T4_6h', 'G4_6h', 'NP68_6h'), 
            cols=c('red', 'green3', 'blue', 'grey40'))
dev.off()

pdf(paste0('../data/CD69_protein_APL2_spikenorm_filt05_violinplot.pdf'), height=2, width=2)
protein_violin_plot('CD69', sceset=sce2, 
            split_facs=c('N4_6h', 'T4_6h', 'G4_6h', 'NP68_6h'), 
            cols=c('red', 'green3', 'blue', 'grey40'))
dev.off()

pdf(paste0('../data/CD25_protein_APL1_spikenorm_filt05_violinplot.pdf'), height=2.5, width=2)
protein_violin_plot('CD25', sceset=sce1, 
            split_facs=c('N4_6h', 'T4_6h', 'G4_6h', 'NP68_6h'), 
            cols=c('red', 'green3', 'blue', 'grey40'))
dev.off()

pdf(paste0('../data/CD25_protein_APL2_spikenorm_filt05_violinplot.pdf'), height=2, width=2)
protein_violin_plot('CD25', sceset=sce2, 
            split_facs=c('N4_6h', 'T4_6h', 'G4_6h', 'NP68_6h'), 
            cols=c('red', 'green3', 'blue', 'grey40'))
dev.off()


```

```{r}
sessionInfo()
```