---
title: "CD8_scRNAseq_QC"
author: "Arianne Richard"
date: "March 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '~/Documents/Post_doc/Git/SingleCellAPL2018/scRNAseq/')
```

This script runs through the QC process for two separate batches of single-cell sequencing data. The first batch contained 2 96-well plates of cells and the second contained 4 96-well plates.

We will process each batch separately and save the resulting datasets as SCESet objects.

Each dataset was aligned with Subread and counted using featureCounts, keeping duplicate reads.  We start this processing from the resulting counts table.

Start with the first batch and read in data and annotation.

```{r}
library(gdata)
library(scater)
library(scran)

counts1 <- read.table('data/counts_table_sortdate20160727.txt', header=TRUE)

annotation <- read.table('data/E-MTAB-6051.sdrf.txt', header=TRUE, sep='\t')
annotation$Sample.Names <- unlist(lapply(strsplit(as.character(annotation$Scan.Name), 'H'), function(x){
  x[1]}))
annotation$Sample.Names <- sub('SLX-', 'SLX.', annotation$Sample.Names)
## remove the duplicates that resulted from multiple sequencing of the same files
annotation <- annotation[!duplicated(annotation$Sample.Names),]
rownames(annotation) <- annotation$Sample.Names

```

Turn counts table into a SingleCellExperiment object and add annotations.

```{r}
sce1 <- SingleCellExperiment(assays=list(counts=as.matrix(counts1)))

pdat <- annotation[colnames(sce1),]
pdat <- pdat[,c('Scan.Name', 
                'Characteristics.well.information.',
                'Characteristics.single.cell.well.quality.',
                'Characteristics.CD69.measurement..log10..', 
                'Characteristics.CD25.measurement..log10..', 
                'Characteristics.CD44.measurement..log10..', 
                'Characteristics.CD62L.measurement..log10..',
                'Parameter.Value.sorting.plate.well.', 
                'Date',
                'Parameter.Value.sequencing.plate.well.',
                'Date.1', 
                'Factor.Value.stimulus.', 
                'Factor.Value.time.'
)]

pdat$Sorting.Plate <- unlist(lapply(strsplit(as.character(pdat$Parameter.Value.sorting.plate.well.), ' '), function(x) {paste0(x[c(1:2)], collapse='')}))

stim <- as.character(pdat$Factor.Value.stimulus.)
stim[stim %in% 'OT-I reduced affinity peptide G4 (SIIGFEKL)'] <- 'G4'
stim[stim %in% 'OT-I high affinity peptide N4 (SIINFEKL)'] <- 'N4'
stim[stim %in% 'OT-I non-binding peptide NP68 (ASNENMDAM)'] <- 'NP68'
stim[stim %in% 'OT-I reduced affinity peptide T4 (SIITFEKL)'] <- 'T4'
stim[stim %in% 'not applicable'] <- 'NA'
stim <- paste0(stim, '_', as.character(pdat$Factor.Value.time.), 'h')
pdat$Condition <- stim

pdat <- DataFrame(pdat)
colData(sce1) <- pdat

## must change this to deal with the gtf file!!
## for version that uses counts files generated by featureCount, this would be fine
## for now, we'll load in an output of featureCounts
load('~/Documents/Post_doc/Sync/Experiments/scRNA-seq/20160727_sc_stim_and_sort/sequencing/20160909_FP34_data/20160909_CD8APL_expt1/data/counts_leaveDup.RData')
fdat <- out$annotation
rownames(fdat) <- as.character(out$annotation$GeneID)
fdat <- fdat[rownames(sce1),]
rowData(sce1) <- fdat

```

Now let's look at various QC metrics.  Due to this dataset having been initially analyzed in 2016, we've done this manually instead of using calculateQCMetrics from the scater package. 

```{r}
mapped <- colSums(counts(sce1))
plot(log2(mapped), xlab='sample', ylab='log2(Assigned)') 
mapped <- data.frame(mapped)
mapped$plate <- colData(sce1)$Sorting.Plate
mapped$condition <- colData(sce1)$Condition
mapped$condition <- factor(mapped$condition, levels=c('N4_6h', 'T4_6h', 'G4_6h', 'NP68_6h', 'NA_NAh'))

boxplot(split(log2(mapped$mapped), mapped$plate), ylab='log2(Assigned)')

boxplot(split(log2(mapped$mapped), mapped$condition), ylab='log2(Assigned)')

colData(sce1)$mapped <- mapped$mapped
```

Look at the ERCC spike-ins and mitochondrial genes.

```{r}
isSpike(sce1, 'ERCC') <- grep('ERCC*', rownames(sce1))
ERCC_sum <- colSums(counts(sce1[isSpike(sce1),]))
genic_sum <- colSums(counts(sce1[!isSpike(sce1),]))

plot(log2(ERCC_sum), log2(genic_sum), xlab='log2(ERCC counts)', ylab='log2(Mm gene counts)')
hist(log2(ERCC_sum)-log2(genic_sum), xlab='log-ratio ERCC : Mm gene')

MT_indices <- grep('chrM', rowData(sce1)$Chr)
MT_sum <- colSums(counts(sce1[MT_indices,]))
genic_noMT_sum <- colSums(counts(sce1)[-c(which(isSpike(sce1)), MT_indices),])

plot(log2(MT_sum), log2(genic_noMT_sum), xlab='log2(mitrochondrial gene counts)', ylab='log2(non-mitochondrial Mm gene counts)')
hist(log2(MT_sum)-log2(genic_noMT_sum), xlab='log-ratio MT : non-MT gene')

colData(sce1)$ERCC_sum <- ERCC_sum
colData(sce1)$genic_sum <- genic_sum
colData(sce1)$MT_sum <- MT_sum
colData(sce1)$genic_noMT_sum <- genic_noMT_sum

```

Look at the number of genes expressed per sample.

```{r}
num_genes_exp <- apply(counts(sce1)[!isSpike(sce1), ], 2, function(x) {length(which(x >0))})

plot(num_genes_exp, ylab='number of genes')

num_genes_exp <- data.frame(num_genes_exp)
num_genes_exp$plate <- colData(sce1)$Sorting.Plate
num_genes_exp$condition <- colData(sce1)$Condition
num_genes_exp$condition <- factor(num_genes_exp$condition, levels=c('N4_6h', 'T4_6h', 'G4_6h', 'NP68_6h', 'NA_NAh'))

boxplot(split(num_genes_exp$num_genes_exp, num_genes_exp$plate), ylab='number of genes')

boxplot(split(num_genes_exp$num_genes_exp, num_genes_exp$condition), ylab='number of genes')

colData(sce1)$num_genes <- num_genes_exp$num_genes_exp
```


Now filter the data for outliers to remove bad cells.  We employ a threshold of 3 median absolute deviations for each parameter.  For library size and number of genes detected, we only filter low cells.  For percentages of ERCC spikes and mitochondrial genes, we filter out cells that are high as recommended in the scran vignette. We will additionally filter out cells with outlying high or low log-ratios of ERCC to non ERCC transcripts as several appeared severely outlying in the plot. 

```{r}
libsize.drop <- isOutlier(sce1$mapped, nmads=3, type="lower", log=TRUE)
feature.drop <- isOutlier(sce1$num_genes, nmads=3, type="lower", log=TRUE)

mito.drop <- isOutlier(sce1$MT_sum/sce1$mapped, nmads=3, type="higher")
spike.drop <- isOutlier(sce1$ERCC_sum/sce1$mapped, nmads=3, type="higher")

outliers_ERCC <- isOutlier(log2(sce1$ERCC_sum)-log2(sce1$genic_sum), nmads=3, type="both")

sce1 <- sce1[,!(libsize.drop | feature.drop | mito.drop | spike.drop | outliers_ERCC)]
data.frame(ByLibSize=sum(libsize.drop), ByFeature=sum(feature.drop), ByMito=sum(mito.drop), BySpike=sum(spike.drop), ByAltSpike=sum(outliers_ERCC), Remaining=ncol(sce1))
```

Now we add feature Data from UCSC mm10 ensGene database. I found this to be more complete than the org.Mm.eg.db database.

```{r}
fdat <- rowData(sce1)
name_map <- read.table('annotation_resources/ucsc_mm10_ensGene', 
                       header=TRUE, sep='\t', quote='\"', comment.char="")
name_map2 <- read.table('annotation_resources/ucsc_mm10_ensemblToGeneName', 
                        row.names=1)
rownames(name_map) <- as.character(name_map$name)
name_map$symbol <- name_map2[rownames(name_map),1]
name_map_simp <- name_map[,c('name2', 'symbol')]
## get rid of multiple entries per gene (due to multiple transcripts in the original file)
name_map_simp <- name_map_simp[!duplicated(name_map_simp),]
name_map_simp <- rbind(name_map_simp, data.frame(name2=rownames(sce1)[isSpike(sce1)], symbol=rownames(sce1)[isSpike(sce1)]))
rownames(name_map_simp) <- as.character(name_map_simp$name2)
rownames(fdat) <- as.character(fdat$GeneID)
fdat$symbol <- name_map_simp[rownames(fdat), 'symbol']
rowData(sce1) <- fdat
```

And finally save the resulting SCEset.

```{r}
saveRDS(sce1, file='data/sce_run1.rds')
```

Now clear the slate and run the same process on sequencing run 2.  This second sequencing run was performed in two separate lanes on two separate dates (the same library pool was run both times).

```{r}
keep(list=list('fdat', 'annotation'))

counts2 <- read.table('data/counts_table_sortdate20161026.txt', header=TRUE)

```

Turn counts table into a SingleCellExperiment object and add annotations.

```{r}
sce2 <- SingleCellExperiment(assays=list(counts=as.matrix(counts2)))

pdat <- annotation[colnames(sce2),]
pdat <- pdat[,c('Scan.Name', 
                'Characteristics.well.information.',
                'Characteristics.single.cell.well.quality.',
                'Characteristics.CD69.measurement..log10..', 
                'Characteristics.CD25.measurement..log10..', 
                'Characteristics.CD44.measurement..log10..', 
                'Characteristics.CD62L.measurement..log10..',
                'Parameter.Value.sorting.plate.well.', 
                'Date',
                'Parameter.Value.sequencing.plate.well.',
                'Date.1', 
                'Factor.Value.stimulus.', 
                'Factor.Value.time.'
)]

pdat$Sorting.Plate <- unlist(lapply(strsplit(as.character(pdat$Parameter.Value.sorting.plate.well.), ' '), function(x) {paste0(x[c(1:2)], collapse='')}))

stim <- as.character(pdat$Factor.Value.stimulus.)
stim[stim %in% 'OT-I reduced affinity peptide G4 (SIIGFEKL)'] <- 'G4'
stim[stim %in% 'OT-I high affinity peptide N4 (SIINFEKL)'] <- 'N4'
stim[stim %in% 'OT-I non-binding peptide NP68 (ASNENMDAM)'] <- 'NP68'
stim[stim %in% 'OT-I reduced affinity peptide T4 (SIITFEKL)'] <- 'T4'
stim[stim %in% 'not applicable'] <- 'NA'
stim <- paste0(stim, '_', as.character(pdat$Factor.Value.time.), 'h')
stim[stim %in% c('unstimulated_NAh')] <- 'unstimulated'
pdat$Condition <- stim

pdat <- DataFrame(pdat)
colData(sce2) <- pdat

## and we can re-use the rowData file from before
rowData(sce2) <- fdat

```

Look at the same diagnostic plots as previously.

```{r}
mapped <- colSums(counts(sce2))
plot(log2(mapped), xlab='sample', ylab='log2(Assigned)') 
mapped <- data.frame(mapped)
mapped$plate <- colData(sce2)$Sorting.Plate
mapped$condition <- colData(sce2)$Condition
mapped$condition <- factor(mapped$condition, levels=c('unstimulated', 'N4_1h', 'N4_3h', 'N4_6h', 'T4_6h', 'G4_6h', 'NP68_6h', 'NA_NAh'))

boxplot(split(log2(mapped$mapped), mapped$plate), ylab='log2(Assigned)')

boxplot(split(log2(mapped$mapped), mapped$condition), ylab='log2(Assigned)')

colData(sce2)$mapped <- mapped$mapped

isSpike(sce2, 'ERCC') <- grep('ERCC*', rownames(sce2))
ERCC_sum <- colSums(counts(sce2[isSpike(sce2),]))
genic_sum <- colSums(counts(sce2[!isSpike(sce2),]))

plot(log2(ERCC_sum), log2(genic_sum), xlab='log2(ERCC counts)', ylab='log2(Mm gene counts)')
hist(log2(ERCC_sum)-log2(genic_sum), xlab='log-ratio ERCC : Mm gene')

MT_indices <- grep('chrM', rowData(sce2)$Chr)
MT_sum <- colSums(counts(sce2[MT_indices,]))
genic_noMT_sum <- colSums(counts(sce2)[-c(which(isSpike(sce2)), MT_indices),])

plot(log2(MT_sum), log2(genic_noMT_sum), xlab='log2(mitrochondrial gene counts)', ylab='log2(non-mitochondrial Mm gene counts)')
hist(log2(MT_sum)-log2(genic_noMT_sum), xlab='log-ratio MT : non-MT gene')

colData(sce2)$ERCC_sum <- ERCC_sum
colData(sce2)$genic_sum <- genic_sum
colData(sce2)$MT_sum <- MT_sum
colData(sce2)$genic_noMT_sum <- genic_noMT_sum

## plot number of genes expressed per sample
num_genes_exp <- apply(counts(sce2)[!isSpike(sce2), ], 2, function(x) {length(which(x >0))})

plot(num_genes_exp, ylab='number of genes')

num_genes_exp <- data.frame(num_genes_exp)
num_genes_exp$plate <- colData(sce2)$Sorting.Plate
num_genes_exp$condition <- colData(sce2)$Condition
num_genes_exp$condition <- factor(num_genes_exp$condition, levels=c('unstimulated', 'N4_1h', 'N4_3h', 'N4_6h', 'T4_6h', 'G4_6h', 'NP68_6h', 'NA_NAh'))

boxplot(split(num_genes_exp$num_genes_exp, num_genes_exp$plate), ylab='number of genes')

boxplot(split(num_genes_exp$num_genes_exp, num_genes_exp$condition), ylab='number of genes')

colData(sce2)$num_genes <- num_genes_exp$num_genes_exp

```

And now apply the same 3 median absolute deviation filters to this sequencing data.

```{r}
libsize.drop <- isOutlier(sce2$mapped, nmads=3, type="lower", log=TRUE)
feature.drop <- isOutlier(sce2$num_genes, nmads=3, type="lower", log=TRUE)

mito.drop <- isOutlier(sce2$MT_sum/sce2$mapped, nmads=3, type="higher")
spike.drop <- isOutlier(sce2$ERCC_sum/sce2$mapped, nmads=3, type="higher")

outliers_ERCC <- isOutlier(log2(sce2$ERCC_sum)-log2(sce2$genic_sum), nmads=3, type="both")

## and because we know from the FACS that there was a well with 2 cells, we need to exclude these as well

notonecell <- sce2$Characteristics.well.information. %in% '2 cells'

sce2 <- sce2[,!(libsize.drop | feature.drop | mito.drop | spike.drop | outliers_ERCC | notonecell)]
data.frame(ByLibSize=sum(libsize.drop), ByFeature=sum(feature.drop), ByMito=sum(mito.drop), BySpike=sum(spike.drop), ByAltSpike=sum(outliers_ERCC), Remaining=ncol(sce2))

```


Save the SCESet object.

```{r}
saveRDS(sce2, file='data/sce_run2.rds')
```

Save session information.

```{r}
sessionInfo()
```
